torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -2106,6 +2106,11 @@ def _pre_backward_hook(*unused: Any) -> None:
             # Run ``_pre_backward_hook`` only once per backward pass
             if self._pre_backward_hook_has_run:
                 return
+            # TODO: (Race) If two autograd threads enter this function and see
```

Comment:
Curious, do we have multiple threads per device? I might miss some recent progress on autograd, I recall it was only one thread per device before?

cc @albanD

Reply:
After some local experimentation, I think we were mistaken about the race, so this PR should not be merged. One thread per device makes sense for what I am seeing empirically in my local experiments.

Reply:
Every device has a single thread associated to it indeed.
A "device" is defined as anything that is not cpu/lazy/meta.
So if you're running on cpu, you can actually have a race as two cpu thread can be working on two independent backward calls at the same time.